etc/ha_healthchecker/ha_healthchecker.py

```diff
@@ -0,0 +1,748 @@
+        try:
+            over_time = iso8601.parse_date(
+                node_obj.heartbeat) + datetime.timedelta(
+                seconds=self.heartbeat_timeout)
+        except (iso8601.ParseError, TypeError, ValueError):
+            # The heartbeat is not formatted, this must be the kp on the node
+            # is not finish the first loop. We just return False, as we are in
+            # the initializing process.
+            return False
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def _ping(self, ipaddr):
+        cli = ['ping', '-c1', '-w1', ipaddr]
+        proc = subprocess.Popen(cli, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+        proc.communicate()
+
+        return proc.returncode == 0
+
+    def _parse_isotime(self, timestr):
+        try:
+            return iso8601.parse_date(timestr)
+        except iso8601.ParseError as e:
+            raise ValueError(six.text_type(e))
+        except TypeError as e:
+            raise ValueError(six.text_type(e))
+
+    def _is_alarmed_timeout(self, obj):
+        over_time = self._parse_isotime(
+            obj.alarmed_at) + datetime.timedelta(
+            days=2)
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def _get_service_status(self, service):
+        # gearman is controlled by zuul itself. It's not under systemd.
+        if service == 'gearman':
+            return 'up' if os.path.exists(
+                '/run/gearman/server.pid') else 'down'
+
+        # timer tasks are handled by crontab
+        if service in ['zuul-timer-tasks', 'nodepool-timer-tasks']:
+            service = 'cron'
+
+        cmd = "systemctl status {srvc}".format(srvc=service)
+        try:
+            # 0 means OK
+            # if using status command, 0 means active, non-zero means error
+            # status.
+            subprocess.check_output(cmd.split(), stderr=subprocess.STDOUT)
+            LOG.debug("Run CMD: %(cmd)s", {'cmd': cmd})
+            return 'up'
+        except subprocess.CalledProcessError as e:
+            LOG.error("Failed to check %(srvc)s service: "
+                      "%(err)s %(out)s", {'srvc': service, 'err': e,
+                                          'out': e.output})
+            return 'down'
+
+
+class Refresher(Base):
+    def __init__(self, zk, heartbeat_timeout):
+        super(Refresher, self).__init__(zk, heartbeat_timeout)
+
+    def _local_node_service_process(self, node_obj):
+        service_objs = self.zk.list_services(node_name_filter=node_obj.name)
+        for service_obj in service_objs:
+            self._refresh_service(service_obj, node_obj)
+
+        self._report_heart_beat(node_obj)
+        if node_obj.alarmed and node_obj.role == 'slave':
+            self.zk.update_node(node_obj.name, alarmed=False)
```

Comment:
let's move L151-L152 into L150. Once report heart beat, if `is_alarmed=True`, change it to `False` at the same time.

Reply:
Correct. Done

Reply:
This logic is just to fix the down slave node manually, and re-online it.